gpkit/keydict.py

```diff
@@ -0,0 +1,212 @@
+import numpy as np
+from collections import defaultdict
+from .small_classes import Numbers, Strings
+from .small_scripts import is_sweepvar, veckeyed
+
+
+class KeyDict(dict):
+    """KeyDicts allow storing and accessing the same value with multiple keys
+
+    A KeyDict keeps an internal list of VarKeys as canonical keys,but allow
+    accessing their values with any object whose `key` attribute matches
+    one of those VarKeys, or with strings who match any of the multiple
+    possible string interpretations of each key.
+
+    In addition, if collapse_arrays is True then VarKeys which have a `shape`
+    parameter (indicating they are part of an array) are stored as numpy
+    arrays, and automatically de-indexed when a matching VarKey with a
+    particular `idx` parameter is used as a key.
+
+    By default a KeyDict will regenerate the list of possible key strings
+    for every usage; a KeyDict may instead be "baked" to have a fixed list of
+    keystrings by calling the `bake()` method.
```

Comment:
This is a fantastic docstring. It would be even better if there were a small code example after every paragraph.